### PR TITLE
Persistent connections & unary handlers

### PR DESCRIPTION
**Pull request by deniskamazur**
This PR implements unary (request-response) handlers over persistent daemon connections.

The hivemind part (Python): https://github.com/learning-at-home/hivemind/pull/328

__TODO for future PRs:__

- [ ] Merge `PersistentConnection{Request,Response}` to `PersistentConnectionMessage` (see [discussion](https://github.com/learning-at-home/go-libp2p-daemon/pull/3#discussion_r688585583))

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -146,6 +146,23 @@ func (d *Daemon) handleConn(c net.Conn) {
...
+			d.handleUpgradedConn(r, w)
+			return
+
+		case pb.Request_ADD_UNARY_HANDLER:
+			fallthrough
+		case pb.Request_SEND_RESPONSE_TO_REMOTE:
+			fallthrough
+		case pb.Request_CALL_UNARY:
+			if err := w.WriteMsg(
+				// TODO: return better error message
```
Comment (borzunov) [inline conn.go:159]: Please do todo before merging :)

File: pb/p2pd.proto [hunk 2]
```diff
@@ -4,15 +4,20 @@ package p2pd.pb;
 
 message Request {
   enum Type {
-    IDENTIFY       = 0;
-    CONNECT        = 1;
-    STREAM_OPEN    = 2;
+    IDENTIFY = 0;
```
Comment (borzunov) [inline pb/p2pd.proto:7]: Please exclude formatting changes of the existing messages from the diff

File: go.mod [hunk 3]
```diff
@@ -5,6 +5,7 @@ go 1.15
 require (
 	github.com/davidlazar/go-crypto v0.0.0-20200604182044-b73af7476f6c // indirect
 	github.com/gogo/protobuf v1.3.2
+	github.com/golang/protobuf v1.4.2 // indirect
```
Comment (borzunov) [inline go.mod:8]: Why do we need to include protobuf one more time?

File: go.sum [hunk 4]
```diff
@@ -184,8 +184,9 @@ github.com/google/pprof v0.0.0-20181206194817-3ea8567a2e57/go.mod h1:zfwlbNMJ+OI
 github.com/google/renameio v0.1.0/go.mod h1:KWCgfxg9yswjAJkECMjeO8J8rahYeXnNhOm40UhjYkI=
 github.com/google/uuid v1.0.0/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
 github.com/google/uuid v1.1.1/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
-github.com/google/uuid v1.1.2 h1:EVhdT+1Kseyi1/pUmXKaFxYsDNy9RQYkMWRH68J/W7Y=
 github.com/google/uuid v1.1.2/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
+github.com/google/uuid v1.3.0 h1:t6JiXgmwXMjEs8VusXIJk2BXHsn+wx8BZdTaoZ5fu7I=
+github.com/google/uuid v1.3.0/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
```
Comment (borzunov) [inline go.sum:189]: nit: Is this change necessary? If not, I'd suggest to revert it to keep the diff with the upstream minimal.

File: go.mod [hunk 5]
```diff
@@ -5,6 +5,7 @@ go 1.15
 require (
 	github.com/davidlazar/go-crypto v0.0.0-20200604182044-b73af7476f6c // indirect
 	github.com/gogo/protobuf v1.3.2
+	github.com/google/uuid v1.3.0
```
Comment (borzunov) [inline go.mod:8]: nit: Is this change necessary? If not, I'd suggest to revert it to keep the diff with the upstream minimal.

File: pb/p2pd.proto [hunk 6]
```diff
@@ -3,18 +3,23 @@ syntax = "proto2";
...
+    CONNMANAGER              = 6;
+    DISCONNECT               = 7;      
+    PUBSUB                   = 8;
+
+    PERSISTENT_CONN_UPGRADE  = 9;
+    CALL_UNARY              = 10;
+    ADD_UNARY_HANDLER       = 11;
+    SEND_RESPONSE_TO_REMOTE = 12;
 
+  }
```
Comment (borzunov) [inline pb/p2pd.proto:22]: nit: Make formatting more neat:

```suggestion
  }
  
```

File: pb/p2pd.proto [hunk 7]
```diff
@@ -26,6 +31,28 @@ message Request {
   optional PSRequest pubsub = 8;
 }
 
+// Persistent connection request
+message PCRequest {
+  required bytes callId = 1;
+
+  oneof message {
+    AddUnaryHandlerRequest addUnaryHandler = 2;
+    CallUnaryRequest  callUnary = 3;
```
Comment (borzunov) [inline pb/p2pd.proto:40]: nit: Make formatting more neat:

```suggestion
    CallUnaryRequest callUnary = 3;
```

File: pb/p2pd.proto [hunk 8]
```diff
@@ -4,15 +4,20 @@ package p2pd.pb;
...
+    DHT                      = 4;
+    LIST_PEERS               = 5;
+    CONNMANAGER              = 6;
+    DISCONNECT               = 7;      
+    PUBSUB                   = 8;
+
+    PERSISTENT_CONN_UPGRADE  = 9;
+    CALL_UNARY              = 10;
+    ADD_UNARY_HANDLER       = 11;
+    SEND_RESPONSE_TO_REMOTE = 12;
```
Comment (borzunov) [inline pb/p2pd.proto:20]:
```suggestion
    RETURN_UNARY_RESPONSE = 12;
```

File: pb/p2pd.proto [hunk 9]
```diff
@@ -155,4 +185,26 @@ message PSMessage {
...
+}
+
+message CallUnaryRequest {
+  required bytes peer = 1;
+  required string proto = 2;
+  required bytes data = 3;
+}
+
+message CallUnaryResponse {
+  optional bytes result = 1;
```
Comment (borzunov) [inline pb/p2pd.proto:197]: Can we use `oneof` here as well?

File: pb/p2pd.proto (hunk 8 above)
Comment (deniskamazur) [reply pb/p2pd.proto:20]: See https://github.com/learning-at-home/hivemind/pull/328

File: go.sum (hunk 4 above)
Comment (deniskamazur) [reply go.sum:189]: Yeah, we need both of them. As long as go check is ok with those lines, I wouldn't bother removing them.

File: persistent_stream.go [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+		callID, err := uuid.FromBytes(req.CallId)
+		if err != nil {
+			log.Debugw("bad call id in p2p handler", "error", err)
+			return
+		}
+
+		// create response channel
+		rc := make(chan *pb.PersistentConnectionRequest)
+		d.responseWaiters.Store(callID, rc)
```
Comment (borzunov) [inline persistent_stream.go:240]:
```suggestion
		responseChan := make(chan *pb.PersistentConnectionRequest)
		d.responseWaiters.Store(callID, responseChan)
```

I suggest to avoid the obvious comments and the comments that can be replaced with a proper variable names. The reason is that comments require time to maintain and can easily become unsynchronized with the actual code, while the variable names usually can't.

File: daemon.go [hunk 11]
```diff
@@ -39,12 +39,21 @@ type Daemon struct {
...
 	// closed is set when the daemon is shutting down
 	closed bool
+
+	// used to check if a handler for a given protocol is
+	// already registered
+	registeredUnaryProtocols map[protocol.ID]bool
+	// maps call id's to wait channels
+	// callId (int64) -> channel to the go routine waiting for client response (chan *pb.Request)
+	responseWaiters sync.Map
+	cancelUnary     sync.Map
```
Comment (borzunov) [inline daemon.go:49]: Please clarify the key/value types for this `sync.Map` as well.

File: p2pclient/p2pclient.go [hunk 12]
```diff
@@ -26,15 +27,23 @@ type Client struct {
 
 	mhandlers sync.Mutex
 	handlers  map[string]StreamHandlerFunc
+
+	openPersistentConn sync.Once
+	// persistent connection writer and reader
+	pConnWriter ggio.WriteCloser
```
Comment (borzunov) [inline p2pclient/p2pclient.go:33]:
```suggestion
	persistentConnWriter ggio.WriteCloser
```

The comment is (1) not synchronized (there is no reader here) and (2) can be replaced with the proper variable name.

File: p2pclient/unary_handlers.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+	select {
+	case done <- struct{}{}:
+		return result.GetResponse(), nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// Cancel cancelles a request w/ given id
```
Comment (borzunov) [inline p2pclient/unary_handlers.go:218]:
```suggestion
```

Obvious comment.

File: p2pclient/unary_handlers.go [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+	response := <-rc.(PersistentConnectionResponseFuture)
+	if dErr := response.GetDaemonError(); dErr != nil {
+		return nil, newDaemonError(dErr)
+	}
+
+	return response, nil
+}
+
+// AddUnaryHandler registers unary handlers to daemon
```
Comment (borzunov) [inline p2pclient/unary_handlers.go:123]:
```suggestion
```

Obvious comment.

**Review by borzunov — CHANGES_REQUESTED** [review]
I have not read much yet, but left the bunch of change requests.

Please apply this to all other comments in the code where appropriate.

File: persistent_stream.go [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+					CallId: callID[:],
+					Message: &pb.PersistentConnectionResponse_Cancel{
+						Cancel: &pb.Cancel{},
+					},
+				},
+			); err != nil {
+				log.Debugw("failed to write to client", "error", err)
+				return
+			}
+			return
```
Comment (borzunov) [inline persistent_stream.go:261]:
```suggestion
			}
```

This can be simplified.

**Review by borzunov — CHANGES_REQUESTED** [review]

File: persistent_stream.go [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+			case <-ctx.Done():
+			}
+		}
+	}()
+
+	return semaphor
+}
+
+// getPersistentStreamHandler returns a lib-p2p stream handler tied to a
+// given persistent client stream
```
[the remark this answers is not in the document]
Comment (deniskamazur) [reply persistent_stream.go:218]: Docstrings in go code begin with the name of the entity they are documenting. Like [here](https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/fmt/print.go;l=230).

File: p2pclient/unary_handlers.go [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+	return c.getPersistentWriter().WriteMsg(
+		&pb.PersistentConnectionRequest{
+			CallId: callID[:],
+			Message: &pb.PersistentConnectionRequest_Cancel{
+				Cancel: &pb.Cancel{},
+			},
+		},
+	)
+}
+
```
Comment (borzunov) [inline p2pclient/unary_handlers.go:221]:
```suggestion
```

Please either remove this function or call it inside `CallUnaryHandler()` (currently, there is duplicated code).

I prefer the first option since this function is unlikely to be called from the outside, since we don't expose call IDs to the user.

File: p2pclient/unary_handlers.go [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+			},
+		},
+	)
+
+	if _, err := c.getResponse(callID); err != nil {
+		return err
+	}
+
+	c.mhandlers.Lock()
+	c.unaryHandlers[proto] = handler
```
Comment (borzunov) [inline p2pclient/unary_handlers.go:138]: Can we make it a `SyncMap`?

File: p2pclient/unary_handlers.go [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+		r.ReadMsg(&resp)
+
+		callID, err := uuid.FromBytes(resp.CallId)
+		if err != nil {
+			log.Debugw("received response with bad call id:", "error", err)
+			continue
+		}
+
+		switch resp.Message.(type) {
+		case *pb.PersistentConnectionResponse_RequestHandling:
```
Comment (borzunov) [inline p2pclient/unary_handlers.go:58]: It is misleading to have a response type that actually corresponds to a request :)

I'd suggest to combine `PersistentConnectionRequest` and `PersistentConnectionResponse` to a single `PersistentConnectionMessage` type (they are already very similar) and send protobufs of this type both ways.

File: test/unary_handler_test.go [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	time.Sleep(time.Second * 3)
+
+	if err := c2.AddUnaryHandler(proto, sqrtHandler); err != nil {
+		t.Fatal("closing client 1 should have cleaned up the proto list", err)
+	}
+}
+
+func sqrtHandler(ctx context.Context, data []byte) ([]byte, error) {
```
Comment (borzunov) [inline test/unary_handler_test.go:154]: The name of this handler does not match its code.

File: test/unary_handler_test.go [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+	defer cancel()
+
+	_, err = p2.CallUnaryHandler(ctx, peer1ID, proto, []byte("hi"))
+	if err == nil {
+		t.Fatal("expected error")
+	}
+}
+
+func TestAddUnaryHandler(t *testing.T) {
+	// create a singe daemon and connect two client to it
```
Comment (borzunov) [inline test/unary_handler_test.go:114]:
```suggestion
	// create a single daemon and connect two clients to it
```

File: test/unary_handler_test.go [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+	if err := c2.AddUnaryHandler(proto, sqrtHandler); err == nil {
+		t.Fatal("adding second unary handler with same name should have returned error")
+	}
+
+	if err := c1.Close(); err != nil {
+		t.Fatal("closing client 1 should not have returned an error")
+	}
+
+	time.Sleep(time.Second * 3)
```
Comment (borzunov) [inline test/unary_handler_test.go:147]:
```suggestion
	time.Sleep(time.Second)
```

If possible, let's limit the delays by 1 sec to save the CI time :)

File: test/unary_handler_test.go [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+
+	peer1ID, peer1Addrs, err := p1.Identify()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := p2.Connect(peer1ID, peer1Addrs); err != nil {
+		t.Fatal(err)
+	}
+
+	var proto protocol.ID = "sqrt"
```
Comment (borzunov) [inline test/unary_handler_test.go:99]: Please change this name since the handler does not actually calculate sqrt.

File: test/unary_handler_test.go [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+	if err := p1.AddUnaryHandler(proto, sqrtHandler); err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	_, err = p2.CallUnaryHandler(ctx, peer1ID, proto, []byte("hi"))
+	if err == nil {
+		t.Fatal("expected error")
```
Comment (borzunov) [inline test/unary_handler_test.go:109]:
```suggestion
		t.Fatal("handler is expected to be cancelled but finished successfully")
```

File: test/unary_handler_test.go [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+	time.Sleep(time.Second * 3)
+
+	if err := c2.AddUnaryHandler(proto, sqrtHandler); err != nil {
+		t.Fatal("closing client 1 should have cleaned up the proto list", err)
+	}
+}
+
+func sqrtHandler(ctx context.Context, data []byte) ([]byte, error) {
+	time.Sleep(time.Second * 30)
```
Comment (borzunov) [inline test/unary_handler_test.go:155]:
```suggestion
	time.Sleep(time.Second * 5)
```

If new changes break cancellation, we don't want tests to freeze for 30 secs before reporting failure :)

File: test/unary_handler_test.go [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+	)
+
+	t.Run(
+		"test correct request",
+		func(t *testing.T) {
+			reply, err := p2.CallUnaryHandler(context.Background(), peer1ID, proto, float64Bytes(64))
+			if err != nil {
+				t.Fatal(err)
+			}
+			result := float64FromBytes(reply)
```
Comment (borzunov) [inline test/unary_handler_test.go:65]: Here, please assert that `result` is correct.

File: test/unary_handler_test.go [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+		closeClient1()
+		closeClient2()
+
+		closeDaemon()
+
+		dir1Closer()
+		dir2Closer()
+	}()
+
+	var proto protocol.ID = "sqrt"
```
Comment (borzunov) [inline test/unary_handler_test.go:133]: Please change this name since the handler does not actually calculate sqrt.

**Review by borzunov — CHANGES_REQUESTED** [review]
I've finished reading the PR. This is likely to be the last batch of change requests from me :)

File: p2pclient/unary_handlers.go (hunk 19 above)
Comment (deniskamazur) [reply p2pclient/unary_handlers.go:58]: I like the suggestion, but I would rather do it in a separate PR, as this would require also updating some code in already approved PR's in the python client.

Comment (borzunov) [reply p2pclient/unary_handlers.go:58]: Okay, I will add a todo item to the description :)

File: internal/utils/safe_writer.go [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+func (sw *safeWriter) WriteMsg(msg proto.Message) error {
+	sw.m.Lock()
+	defer sw.m.Unlock()
+	return sw.w.WriteMsg(msg)
+}
+
+func (sw *safeWriter) Close() error {
+	sw.m.Lock()
+	defer sw.m.Unlock()
+
```
Comment (borzunov) [inline internal/utils/safe_writer.go:28]:
```suggestion
```

To make it similar to the previous function.

**Review by borzunov — APPROVED** [review]

**Review by mryab — APPROVED** [review]
